src/daemon/distribution.c

```diff
@@ -0,0 +1,403 @@
+  if (num_buckets == 0 || size <= 0) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  distribution_t *d = calloc(1, sizeof(distribution_t));
+
+  if (d == NULL) {
+    return NULL;
+  }
+
+  d->buckets = bucket_new_linear(num_buckets, size);
+
+  if (d->buckets == NULL) {
+    free(d);
+    return NULL;
+  }
+
+  d->num_buckets = num_buckets;
+  pthread_mutex_init(&d->mutex, NULL);
+
+  return d;
+}
+
+distribution_t *distribution_new_exponential(size_t num_buckets, double base,
+                                             double factor) {
+  if (num_buckets == 0 || base <= 1 || factor <= 0) {
+    errno = EINVAL;
+    return NULL;
+  }
+
+  distribution_t *d = calloc(1, sizeof(distribution_t));
+
+  if (d == NULL) {
+    return NULL;
+  }
+
+  /* as in distribution_new_linear: it would be nice to check if base
+   * and factor are greater than zero, for consideration: one of them also
+   * greater than one */
+  d->buckets = bucket_new_exponential(num_buckets, base, factor);
+
+  if (d->buckets == NULL) {
+    free(d);
+    return NULL;
+  }
+
+  d->num_buckets = num_buckets;
+  pthread_mutex_init(&d->mutex, NULL);
+
+  return d;
+}
+
+distribution_t *distribution_new_custom(size_t num_boundaries,
+                                        double *custom_buckets_boundaries) {
+  distribution_t *d = calloc(1, sizeof(distribution_t));
+
+  if (d == NULL) {
+    return NULL;
+  }
+
+  d->buckets = bucket_new_custom(num_boundaries, custom_buckets_boundaries);
+
+  if (d->buckets == NULL) {
+    free(d);
+    return NULL;
+  }
+
+  d->num_buckets = num_boundaries + 1;
+  pthread_mutex_init(&d->mutex, NULL);
+
+  return d;
+}
+
+static void bucket_update(bucket_t *buckets, size_t num_buckets, double gauge) {
+  int idx = (int)num_buckets - 1;
+
+  while (idx >= 0 && buckets[idx].max_boundary > gauge) {
+    buckets[idx].counter++;
+    idx--;
+  }
+}
+
+int distribution_update(distribution_t *d, double gauge) {
+  if (d == NULL || gauge < 0) {
+    errno = EINVAL;
+    return EXIT_FAILURE;
+  }
+
+  pthread_mutex_lock(&d->mutex);
+
+  bucket_update(d->buckets, d->num_buckets, gauge);
+
+  d->sum_gauges += gauge;
+  pthread_mutex_unlock(&d->mutex);
+
+  return EXIT_SUCCESS;
+}
+
+static double find_percentile(bucket_t *buckets, size_t num_buckets,
+                              uint64_t quantity) {
+  size_t left = 0;
+  size_t right = num_buckets - 1;
+  size_t middle;
+
+  while (left < right) {
+    middle = (left + right) / 2;
+
+    if (buckets[middle].counter >= quantity) {
+      right = middle;
+    } else {
+      left = middle + 1;
+    }
+  }
+
+  return buckets[left].max_boundary;
+}
+
+double distribution_percentile(distribution_t *d, double percent) {
+  if (d == NULL || percent > 100.0 || percent < 0) {
+    errno = EINVAL;
+    return NAN;
+  }
+
+  pthread_mutex_lock(&d->mutex);
+
+  uint64_t quantity = (uint64_t)(
+      (percent / 100.0) * (double)d->buckets[d->num_buckets - 1].counter);
```

Comment:
BTW, what would you return if distribution was created but has no updates yet? I'd argue that you should return NAN

Reply:
Fixed, thanks!